File: graph/number_of_islands.py
from typing import List

# ====================== DFS  ====================== #
# mark visited in-place
# O(m*n) time
# O(m*n) space: entirely from recursion call stack
    # In the worst case, the entire grid is land ('1') arranged so that the DFS traverses it as one long non-branching path — for example a snake-like shape where each cell leads to exactly one unvisited neighbor.
class Solution:
    def numIslands(self, grid: List[str]) -> int:
        m, n = len(grid), len(grid[0])
        def dfs(x: int, y: int) -> None:
            if not (0 <= x < m and 0 <= y < n) or grid[x][y] != '1':
                return
            grid[x][y] = '#'
            dfs(x + 1, y)
            dfs(x - 1, y)
            dfs(x, y + 1)
            dfs(x, y - 1)
        islands = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    islands += 1
                    dfs(i, j)
        return islands

# mark visited in-place
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        islands = 0
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def dfs(x, y):
            grid[x][y] = '#'
            for (dx, dy) in directions:
                adjx = x + dx
                adjy = y + dy
                if (0 <= adjx < m) and (0 <= adjy < n) and (grid[adjx][adjy] == '1'):
                    dfs(adjx, adjy)
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    islands += 1
                    dfs(i, j)
        return islands

# use hash set to mark visited
# adds another O(m*n) for the set (not recommended)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        islands = 0
        visited = set()
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def dfs(x, y):
            visited.add((x, y))
            for (dx, dy) in directions:
                adjx = x + dx
                adjy = y + dy
                if (0 <= adjx < m) and (0 <= adjy < n) and (grid[adjx][adjy] == '1') and ((adjx, adjy) not in visited):
                    dfs(adjx, adjy)
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' and (i, j) not in visited:
                    islands += 1
                    dfs(i, j)
        return islands



# ====================== BFS  ====================== #
# O(m*n) time
# O(min(m*n)) space: why?
from collections import deque
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        islands = 0
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        def bfs(i, j):
            grid[i][j] = '#' 
            q = deque([])
            q.append((i, j))
            while q:
                x, y = q.popleft()
                for (dx, dy) in directions:
                    adjx = x + dx
                    adjy = y + dy
                    if (0 <= adjx < m) and (0 <= adjy < n) and (grid[adjx][adjy] == '1'):
                        grid[adjx][adjy] = '#'
                        q.append((adjx, adjy))

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    islands += 1
                    bfs(i, j)
        return islands

File: graph/number_of_islands.cpp
// BFS
// use some modern C++ features
// O(m*n) time
// O(min(m,n)) space
class Solution {
private:
    // Use static to avoid re-allocation on stack every call
    // array for 4 adjacent directions: left, right, up, down
    static constexpr int directions[4][2] = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};
    int m, n;

    void bfs(std::vector<vector<char>>& grid, int i, int j) {
        std::queue<std::pair<int, int>> q; // queue to visit cells in BFS
        q.push({i,j});
        grid[i][j] = '#'; // must mark the starting cell as visited to prevent revisiting
        
        while (!q.empty()) {
            auto [x, y] = q.front(); // structured binding, ok since C++ 17
            q.pop();
            // check 4 directions
            for (const auto& direction : directions) {
                // previous method: for (auto direction : directions) is a copy
                int adjx = x + direction[0];
                int adjy = y + direction[1];
                // check boundaries and whether the cell represent a land('1')
                if (adjx >= 0 && adjx < m
                    && adjy >= 0 && adjy < n
                    && grid[adjx][adjy] == '1'
                ) {
                    grid[adjx][adjy] = '#'; // mark as visited
                    q.push({adjx, adjy});
                }
            }
        }
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty() || grid[0].empty()) return 0;
        // number of rows and columns
        m = grid.size();
        n = grid[0].size();
        int islands = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == '1') {
                    ++islands;
                    bfs(grid, i, j);
                }
            }
        }
        return islands;
        
    }
};